herbert_core/utilities/maths/sawtooth_iarray.m

```diff
@@ -1,38 +1,37 @@
 function ivout = sawtooth_iarray (n)
-% Create array of successive runs [1,2,3..n(1),1,2,3...n(2),...]'
+% Create column vector [1;2;3..n(1); 1;2;3...n(2); ...] from integer array n
 %
 %   >> ivout = sawtooth_iarray (n)
 %
+%
 % Input:
 % ------
-%   n       List of length of each sawtooth section (integer)
-%          (Zeros correspond to zero length section i.e. they are
-%           effectively ignored).
-%           Must have all(n(:)>=0).
+%   n       Array with the lengths of each section of the output column vector.
+%           - Each section is created as the column vector (1:n(i))'
+%           - The elements of n should have integer values greater or equal to zero.
+%           - Zeros correspond to zero length sections i.e. they are ignored).
+%             This is consistent with Matlab which constructs M:N as an empty
+%             array if N < M.
 %
 % Output:
 % -------
-%   ivout   Output array: column vector
-%               ivout=[1:n(1),1:n(2),...]'
-%
-% NOTE: This is designed for integer array n only, as it assumes that
-%       there are no rounding errors on addition.
+%   ivout   Output array: always a column vector with length sum(n(:)).
+%           It is constructed as:
+%               ivout [1;2;3..n(1); 1;2;3...n(2); ...] 
 
-% Original author: T.G.Perring
-%
 
+n = n(:);
+if any(n<0) || any(round(n)~=n)
+    error('HERBERT:sawtooth_iarray:invalid_argument',...
+        'The elements of the input argument must all be nonnegative integer values')
+end
 
-nn0=n(n>0);
-nn0=nn0(:);
-if ~isempty(nn0)
-    nend=cumsum(nn0);
-    tmp=ones(nend(end),1);
-    ix=1+nend(1:end-1);
-    if ~isempty(ix)
-        tmp(ix)=tmp(ix)-double(nn0(1:end-1));
-    end
-    ivout=cumsum(tmp);
+n = n(n>0);
+if isempty(n)
+    ivout = zeros(0,1);
+elseif isscalar(n)
+    ivout = (1:n)';
 else
-    ivout=[];
+    offset = cumsum([0;n]);
+    ivout = (1:offset(end))' - repelem(offset(1:end-1), n);
```

Comment:
I have stared at this blankly for some time. I'm sure it works, but a line of explanation would be good

Reply:
I've added an explanation of the algorithm. Note that I branch on the case of empty and scalar n to avoid the overhead of calling repelem for these simple cases. It keeps the algorithm optimal.